### ISSUE TITLE
TODO: Make "deterministic Trap" and "nondeterministic Trap" types, and allow cache to store the former

### ISSUE THREAD

**Issue by keithw** [fix-project/fix#63]

**Comment by Akshay-Srivatsan** [comment]
Do we want to allow user programs to purposefully trigger deterministic traps, e.g. to signal invalid inputs?  Or are deterministic traps specifically for handling traps generated within API calls?  It seems like there wouldn't be much harm in having a `fix_trap` API call, since the program could just call `attach_blob` on a tree or something to trigger a trap anyway.

**Comment by keithw** [comment]
Every Wasm trap (except for stack overflow) is a deterministic trap, so we can't really stop them if they want to... The cleanest way for a Wasm function to trap is just to execute the "unreachable" instruction. So I don't think we need an API call for this -- running "unreachable" is already pretty easy.

**Comment by Akshay-Srivatsan** [comment]
I guess the use case I'm thinking of is propagating errors better (e.g., in bootstrap)—just running "unreachable" wouldn't provide a way to specify an error message for debugging purposes.

**Comment by keithw** [comment]
Yeah... I think if we want "good" error handling, it can't just trap. But it also shouldn't be returning a Runnable-tagged error string. Here are some options that maybe sound plausible?

1) There's two procedures, which we could call "trusted-toolchain" and "unwrap".
    - The "trusted-toolchain" runs the whole pipeline (wasm2c+initcomposer->libclang->lld) and returns either
      - if successful, a [trusted-toolchain:Runnable]-tagged ELF Blob
      - otherwise, a [trusted-toolchain:Error]-tagged Blob (or Tree of Blobs?) containing the diagnostic information
      
      This is kind of like a C++ std::expected / Rust Result / Haskell Either?
    - "unwrap" takes a Tag + two Blobs that provide the desired type. If the Tag matches the desired type, it returns it. Otherwise it traps, maybe just after doing an `unsafe-io()` to print out the object (the diagnostic) to be nice?

   To compile something, the user probably runs unwrap(trusted-toolchain(something.wasm)). And if they get a trap and want to see the error message more programmatically, they can always rerun trusted-toolchain(something.wasm) (which is probably cached anyway) to examine the error message.
  
2) The "trusted-toolchain" could **always** return a valid [trusted-toolchain:Runnable]-tagged ELF Blob. If there was an error at any point, it returns a special executable that always returns a constant Blob (or ErrorMessage-tagged Blob) containing the error message.

   Now that I type this one out, it feels a little too cute...
